test/unit/puppetlabs/puppetserver/lockable_pool_test.clj

```diff
@@ -167,54 +207,314 @@
+      @lock-started?
+      (is (thrown? IllegalStateException (.unlock pool)))
+      (deliver unlock? true)
+      @lock-thread
+      (is (true? true)))))
+
 (deftest pool-release-item-test
   (testing (str "releaseItem call with value 'false' does not return item to "
                 "pool but does allow pool to still be lockable")
-    (let [pool (create-populated-pool 1)
-          lock (jruby-testutils/get-lock-from-pool pool)
+    (let [pool (create-populated-pool 2)
           instance (.borrowItem pool)]
-      (is (= 0 (.size pool)))
+      (is (= 1 (.size pool)))
       (.releaseItem pool instance false)
-      (is (= 0 (.size pool)))
+      (.unregister pool instance)
+      (is (= 1 (.size pool)))
       (.lock pool)
-      (is (.isWriteLocked lock))
+      (is (nil? @(future (.borrowItemWithTimeout pool
+                                                 1
+                                                 TimeUnit/MICROSECONDS))))
       (.unlock pool)
-      (is (not (.isWriteLocked  lock)))))
+      (is (not (nil? @(future (.borrowItemWithTimeout pool
+                                                      1
+                                                      TimeUnit/MICROSECONDS)))))
+      (is (true? true))))
   (testing (str "releaseItem call with value 'true' returns item to "
                 "pool and allows pool to still be lockable")
-    (let [pool (create-populated-pool 1)
-          lock (jruby-testutils/get-lock-from-pool pool)
+    (let [pool (create-populated-pool 2)
           instance (.borrowItem pool)]
-      (is (= 0 (.size pool)))
-      (.releaseItem pool instance true)
       (is (= 1 (.size pool)))
+      (.releaseItem pool instance true)
+      (is (= 2 (.size pool)))
       (.lock pool)
-      (is (.isWriteLocked lock))
+      (is (nil? @(future (.borrowItemWithTimeout pool
+                                                 1
+                                                 TimeUnit/MICROSECONDS))))
       (.unlock pool)
-      (is (not (.isWriteLocked  lock))))))
+      (is (not (nil? @(future (.borrowItemWithTimeout pool
+                                                      1
+                                                      TimeUnit/MICROSECONDS)))))
+      (is (true? true)))))
+
+(deftest pool-borrow-blocks-borrow-when-pool-empty
+  (testing "borrow from pool blocks while the pool is empty"
+    (let [pool (create-populated-pool 1)
+          item (.borrowItem pool)
+          borrow-thread-started? (promise)
+          borrow-thread-borrowed? (promise)
+          borrow-thread (future (deliver borrow-thread-started? true)
+                                (let [instance (.borrowItem pool)]
+                                  (deliver borrow-thread-borrowed? true)
+                                  (.releaseItem pool instance)))]
+      @borrow-thread-started?
+      (is (not (realized? borrow-thread-borrowed?)))
+      (.releaseItem pool item)
+      @borrow-thread
+      (is (true? true)))))
```

Comment:
Yeah, I don't think this found a problem with my implementation.  Just some additional validation of the overall behavior.